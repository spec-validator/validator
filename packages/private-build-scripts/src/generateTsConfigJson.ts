import { flatMap } from '@spec-validator/utils/utils'
import { Task } from 'just-task'
import { getGraph, getProjectsPathsInBuildOrder } from './buildOrder'
import { write } from './readAndWrite'
import getRelativePath from './relPath'

export const DIST = 'dist'

const generateRootConfig = (baseTsConfig: string): void =>
  write('tsconfig.build.json', {
    'extends': baseTsConfig,
    'compilerOptions': {
      'noEmit': true,
    },
    'references': getProjectsPathsInBuildOrder().map(it => ({
      'path': `${it}/tsconfig.build.json`,
    })),
  })

const generateProjectConfigs = (baseTsConfig: string): void =>
  Object.entries(getGraph()).forEach(([parent, children]) => {
    write(`${getRelativePath(parent)}/tsconfig.build.json`, {
      'extends': baseTsConfig,
      'compilerOptions': {
        'baseUrl': './',
        'composite': true,
        'noEmit': false,
        'rootDir': './src',
        'outDir': `./${DIST}`,
        'paths': Object.fromEntries(flatMap(
          children.map(child => [
            getRelativePath(parent, child),
            child,
          ]).map(([rel, child]) => [
            [child, [`${rel}/src/index.ts`]],
            [`${child}/*`, [`${rel}/src/*`]],
          ]))),
      },
      'include': ['src/**/*.ts'],
      'exclude': ['src/**/*.spec.ts', 'src/**/*.test.ts'],
      'references': children.map(child => (
        { 'path': `${getRelativePath(parent, child)}/tsconfig.build.json` }
      )),
    })
  })

export default (baseTsConfig: string): Task => async () => {
  generateProjectConfigs(baseTsConfig)
  generateRootConfig(baseTsConfig)
}
