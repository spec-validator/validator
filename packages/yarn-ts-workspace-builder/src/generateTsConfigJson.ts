import path from 'path'

import { flatMap } from '@spec-validator/utils/utils'
import { getGraph, getProjectsPathsInBuildOrder } from './buildOrder'
import { write } from './readAndWrite'
import getRelativePath from './relPath'
import { findFolderWith } from '@spec-validator/cli'

export const DIST = 'dist'

const generateRootConfig = (baseTsConfig: string): void =>
  write('tsconfig.build.json', {
    'extends': baseTsConfig,
    'compilerOptions': {
      'noEmit': true,
    },
    'references': getProjectsPathsInBuildOrder().map(it => ({
      'path': `${it}/tsconfig.build.json`,
    })),
  })

const generateProjectConfigs = (baseTsConfig: string): void =>
  Object.entries(getGraph()).forEach(([parent, children]) => {
    write(`${getRelativePath(parent)}/tsconfig.build.json`, {
      'extends': baseTsConfig,
      'compilerOptions': {
        'baseUrl': './',
        'composite': true,
        'noEmit': false,
        'rootDir': './src',
        'outDir': `./${DIST}`,
        'paths': Object.fromEntries(flatMap(
          children.map(child => [
            getRelativePath(parent, child),
            child,
          ]).map(([rel, child]) => [
            [child, [`${rel}/src/index.ts`]],
            [`${child}/*`, [`${rel}/src/*`]],
          ]))),
      },
      'include': ['src/**/*.ts'],
      'exclude': ['src/**/*.spec.ts', 'src/**/*.test.ts'],
      'references': children.map(child => (
        { 'path': `${getRelativePath(parent, child)}/tsconfig.build.json` }
      )),
    })
  })

const getDefaultTsConfig = (): string => path.join(findFolderWith('.git', process.cwd()), 'tsconfig.json')

export default (baseTsConfig?: string): void => {
  baseTsConfig = baseTsConfig || getDefaultTsConfig()
  generateProjectConfigs(baseTsConfig)
  generateRootConfig(baseTsConfig)
}
