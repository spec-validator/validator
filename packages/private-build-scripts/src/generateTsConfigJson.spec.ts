import generateTsConfigJson from './generateTsConfigJson'

const write = jest.fn()

jest.mock('./relPath', () => ({
  default: (parent: string, child: string) =>
    child ? `../../${child}` : `./${parent}`,
  __esModule: true,
}))

jest.mock('./readAndWrite', () => ({
  write: (path: string, data: any) => write(path, data),
  __esModule: true,
}))

jest.mock('./buildOrder', () => ({
  getProjectsPathsInBuildOrder: () => ['pkg1', 'pkg2'],
  getGraph: () => ({
    'pkg1': ['pkg2'],
    'pkg2': [],
  }),
  __esModule: true,
}))

test('generateTsConfigJson', async () => {
  const run = generateTsConfigJson('./root/tsconfig.json') as any
  await run()
  expect(write.mock.calls).toMatchSnapshot()
})
