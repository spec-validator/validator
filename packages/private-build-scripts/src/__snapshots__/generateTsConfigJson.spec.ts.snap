// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateTsConfigJson 1`] = `
Array [
  Array [
    "./pkg1/tsconfig.build.json",
    Object {
      "compilerOptions": Object {
        "baseUrl": "./",
        "composite": true,
        "noEmit": false,
        "outDir": "./dist",
        "paths": Object {
          "pkg2": Array [
            "../../pkg2/src/index.ts",
          ],
          "pkg2/*": Array [
            "../../pkg2/src/*",
          ],
        },
        "rootDir": "./src",
      },
      "exclude": Array [
        "src/**/*.spec.ts",
        "src/**/*.test.ts",
      ],
      "extends": "./root/tsconfig.json",
      "include": Array [
        "src/**/*.ts",
      ],
      "references": Array [
        Object {
          "path": "../../pkg2/tsconfig.build.json",
        },
      ],
    },
  ],
  Array [
    "./pkg2/tsconfig.build.json",
    Object {
      "compilerOptions": Object {
        "baseUrl": "./",
        "composite": true,
        "noEmit": false,
        "outDir": "./dist",
        "paths": Object {},
        "rootDir": "./src",
      },
      "exclude": Array [
        "src/**/*.spec.ts",
        "src/**/*.test.ts",
      ],
      "extends": "./root/tsconfig.json",
      "include": Array [
        "src/**/*.ts",
      ],
      "references": Array [],
    },
  ],
  Array [
    "tsconfig.build.json",
    Object {
      "compilerOptions": Object {
        "noEmit": true,
      },
      "extends": "./root/tsconfig.json",
      "references": Array [
        Object {
          "path": "pkg1/tsconfig.build.json",
        },
        Object {
          "path": "pkg2/tsconfig.build.json",
        },
      ],
    },
  ],
]
`;
